### PR TITLE
JSDK-2525  create workflow in circleci to build/test twilio-video.js 

### PR DESCRIPTION
**Pull request by makarandp0**
This change adds two circleci workflows for our CI pipeline

1. JSDK-2526: Build_Docker_Images_Workflow ([Example Run](https://circleci.com/workflow-run/7dabee75-c6b8-4afc-a7c8-c6eb061f9296))

This workflow will run nightly and generate docker images with os + node + browser flavor.
That's total 6 images [firefox, chrome] x [stable, unstable, beta]. The job compares the browser versions on newly generated images and if its different that old version it uploads the images to [twilio docker hub](https://cloud.docker.com/u/twilio/repository/docker/twilio/twilio-video-browsers).

2. JSDK-2525: Pull_Request_Workflow ([Example Run](https://circleci.com/workflow-run/056b91a9-7966-41a6-8a1d-b11931677924)) 
This will run on every commit on PR, and run all the jobs that our travis-CI pipeline used to run.
+ integration tests: [chrome, firefox] x [stable, unstable, beta] x [group, p2p]
+ network tests:  [chrome, firefox] x [stable] x [group, p2p]
+ unit tests: 
+ framework tests

it uses the docker images generated by workflow # 1 for all these jobs. 

Immediate benefit is quick and easy access to test results. 
- Every failed job shows [failed test prominently](https://circleci.com/gh/twilio/twilio-video.js/297) at the top
- Every job also stores [full test results](https://307-46595751-gh.circle-artifacts.com/0/home/circleci/project/logs/DO-true--BR-chrome-BV-stable-TO-peer-to-peer-FILE-reconnection.html)
- You can see [coverage results](https://288-46595751-gh.circle-artifacts.com/0/home/circleci/project/coverage/lcov-report/index.html) from unit test runs

Few more changes are needed to completely transition to CI, We will continue to run Travis and CircleCI side by side in the meanwhile.

- support for 1.x branch CI job. [JSDK-2529](https://issues.corp.twilio.com/browse/JSDK-2529)
- changes to release-tool to use this workflow. [JSDK-2528](https://issues.corp.twilio.com/browse/JSDK-2528)

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by charliesantos** [comment]
Looking good @makarandp0 . A couple of questions:

- What would be the side effect or disadvantage if we combine chrome and firefox into one image? So we would have only 3 images instead of 6 - chrome & firefox beta, stable, and stable.

- Does CircleCI support invoking from command line or via APIs? I'm not sure how video does it, but we trigger our release builds from command line.

**Comment by makarandp0** [comment]
Thanks @charliesantos for your comments.
Re: disadvantage of combining chrome/firefox - I don't see much difference in these approaches. The only disadvantage I can think of are 1) If firefox / chrome dependendencies were to have any mismatch it would be harder to get it working. 2) Keeping them seperate may be helpful as we  add more browsers to the mix - say edge, safari etc. As they may not have same 'dev', 'stable' 'beta' flavors. what if a browser has just 2, or 4 flavors.

2. CircleCI does have API to invoke jobs. I am going to look more into it soon as part of JSDK-2528. In case of Video, we use release tool to release our builds, which creates a plan and executes it on the CI box (not local machine). I thought voice uses the same tool. Let me sync up to discuss more about this.

**Comment by charliesantos** [comment]
> Thanks @charliesantos for your comments.
> Re: disadvantage of combining chrome/firefox - I don't see much difference in these approaches. The only disadvantage I can think of are 1) If firefox / chrome dependendencies were to have any mismatch it would be harder to get it working. 2) Keeping them seperate may be helpful as we add more browsers to the mix - say edge, safari etc. As they may not have same 'dev', 'stable' 'beta' flavors. what if a browser has just 2, or 4 flavors.
> 
> 1. CircleCI does have API to invoke jobs. I am going to look more into it soon as part of JSDK-2528. In case of Video, we use release tool to release our builds, which creates a plan and executes it on the CI box (not local machine). I thought voice uses the same tool. Let me sync up to discuss more about this.

@makarandp0 
1) Sounds good. Creating separate images per browser makes more sense.
2) We are doing the same thing - we are using release tool to create a plan and execute it on the CI box. Just wanted to double check that this will still be supported.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,11 +1,328 @@
...
+  version: 2
+  Build_Docker_Images_Workflow:
+    triggers:
+      - schedule:
+          cron: "0 0 * * *"
+          filters:
+            branches:
+              only:
+                - master
+                - makarand_circleci
```
Comment (makarandp0) [inline .circleci/config.yml]: TODO: remove `makarand_circleci` branch from here, before merging.

File: scripts/karma.js [hunk 2]
```diff
@@ -28,20 +28,57 @@ function getTestPaths(path) {
   return [path];
 }
 
-const files = getTestPaths(isDocker ? dockerIntegrationTests : integrationTests);
+function filterTests(paths) {
```
Comment (makarandp0) [inline scripts/karma.js]: This is a quick way to enable distributing test run across machines. I was planning to experiment this, but wanted to get some input on this.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,11 +1,328 @@
...
+          name: "framework tests"
+          requires: [BuildComplete]
+      - BuildPassed:
+          requires:
+            - "framework tests"
+            - "firefox stable p2p"
+            - "firefox stable group"
+            - "chrome stable p2p"
+            - "chrome stable group"
+
```
Comment (manjeshbhargav) [inline .circleci/config.yml]: @makarandp0 Nitpick: remove extra line?

File: .circleci/images/firefox/beta/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ENV DISPLAY :99
+RUN printf '#!/bin/sh\nXvfb :99 -screen 0 1280x1024x24 &\nexec "$@"\n' > /tmp/entrypoint \
+  && chmod +x /tmp/entrypoint \
+        && sudo mv /tmp/entrypoint /docker-entrypoint.sh
+
+# ensure that the build agent doesn't override the entrypoint
+LABEL com.circleci.preserve-entrypoint=true
+
+ENTRYPOINT ["/docker-entrypoint.sh"]
+CMD ["/bin/sh"]
```
Comment (manjeshbhargav) [inline .circleci/images/firefox/beta/Dockerfile]: @makarandp0 Nitpick: add newline?

File: .circleci/images/firefox/stable/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ENV DISPLAY :99
+RUN printf '#!/bin/sh\nXvfb :99 -screen 0 1280x1024x24 &\nexec "$@"\n' > /tmp/entrypoint \
+  && chmod +x /tmp/entrypoint \
+        && sudo mv /tmp/entrypoint /docker-entrypoint.sh
+
+# ensure that the build agent doesn't override the entrypoint
+LABEL com.circleci.preserve-entrypoint=true
+
+ENTRYPOINT ["/docker-entrypoint.sh"]
+CMD ["/bin/sh"]
```
Comment (manjeshbhargav) [inline .circleci/images/firefox/stable/Dockerfile]: @makarandp0 Nitpick: add newline?

File: .circleci/images/firefox/unstable/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ENV DISPLAY :99
+RUN printf '#!/bin/sh\nXvfb :99 -screen 0 1280x1024x24 &\nexec "$@"\n' > /tmp/entrypoint \
+  && chmod +x /tmp/entrypoint \
+        && sudo mv /tmp/entrypoint /docker-entrypoint.sh
+
+# ensure that the build agent doesn't override the entrypoint
+LABEL com.circleci.preserve-entrypoint=true
+
+ENTRYPOINT ["/docker-entrypoint.sh"]
+CMD ["/bin/sh"]
```
Comment (manjeshbhargav) [inline .circleci/images/firefox/unstable/Dockerfile]: @makarandp0 Nitpick: add newline?

File: scripts/karma.js [hunk 7]
```diff
@@ -28,20 +28,57 @@ function getTestPaths(path) {
...
   if (isDocker) {
     console.log('running tests inside docker!');
-    dockerProxy = new DockerProxyServer();
-    dockerProxy.startServer();
+    try {
+      dockerProxy = new DockerProxyServer();
+      await dockerProxy.startServer();
+      console.log('DockerProxyServer started successfully. Network tests may run as part of this run!');
+    } catch (err) {
+      // NOTE: This can happen in CI environment, when we run integration tests inside docker
```
Comment (manjeshbhargav) [inline scripts/karma.js]:
```suggestion
      // NOTE(mpatwardhan): This can happen in CI environment, when we run integration tests inside docker
```

**Comment by manjeshbhargav** [comment]
@makarandp0 One more thing: we should integrate Circle CI to our github repo.

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav - 

> Please consider manually grouping tests to take max advantage of parallel runs.
I will experiment on breaking up tests as we talked to reduce the build times. 

> @makarandp0 One more thing: we should integrate Circle CI to our github repo.
Yes, I think it is integrated with the repro already. But once it goes into master it will show up more clearly. Another reason to ensure that we do not see "red" builds because of unstable tests.

**Pull request merged**